api/pkg/apis/v1alpha1/providers/secret/k8sSecretProvider.go

```diff
@@ -0,0 +1,179 @@
+	}
+	if v, ok := properties["inCluster"]; ok {
+		val := v
+		if val != "" {
+			bVal, err := strconv.ParseBool(val)
+			if err != nil {
+				return ret, v1alpha2.NewCOAError(err, "invalid bool value in the 'inCluster' setting of K8s state provider", v1alpha2.BadConfig)
+			}
+			ret.InCluster = bVal
+		}
+	}
+	return ret, nil
+}
+
+func (s *K8sSecretProvider) InitWithMap(properties map[string]string) error {
+	config, err := K8sSecretProviderConfigFromMap(properties)
+	if err != nil {
+		return err
+	}
+	return s.Init(config)
+}
+
+func (s *K8sSecretProvider) Init(config providers.IProviderConfig) error {
+	_, span := observability.StartSpan("K8s Secret Provider", context.TODO(), &map[string]string{
+		"method": "Init",
+	})
+	var err error = nil
+	defer observ_utils.CloseSpanWithError(span, &err)
+
+	sLog.Debug("  P (K8s Secret): initialize")
+
+	updateConfig, err := toK8sStateProviderConfig(config)
+	if err != nil {
+		sLog.Errorf("  P (K8s Secret): expected KubectlTargetProviderConfig: %+v", err)
+		return err
+	}
+	s.Config = updateConfig
+	var kConfig *rest.Config
+	if s.Config.InCluster {
+		kConfig, err = rest.InClusterConfig()
+	} else {
+		switch s.Config.ConfigType {
+		case "path":
+			if s.Config.ConfigData == "" {
+				if home := homedir.HomeDir(); home != "" {
+					s.Config.ConfigData = filepath.Join(home, ".kube", "config")
+				} else {
+					err = v1alpha2.NewCOAError(nil, "can't locate home direction to read default kubernetes config file, to run in cluster, set inCluster config setting to true", v1alpha2.BadConfig)
+					sLog.Errorf("  P (K8s State): %+v", err)
+					return err
+				}
+			}
+			kConfig, err = clientcmd.BuildConfigFromFlags("", s.Config.ConfigData)
+		case "bytes":
+			if s.Config.ConfigData != "" {
+				kConfig, err = clientcmd.RESTConfigFromKubeConfig([]byte(s.Config.ConfigData))
+				if err != nil {
+					sLog.Errorf("  P (K8s State): %+v", err)
+					return err
+				}
+			} else {
+				err = v1alpha2.NewCOAError(nil, "config data is not supplied", v1alpha2.BadConfig)
+				sLog.Errorf("  P (K8s State): %+v", err)
+				return err
+			}
+		default:
+			err = v1alpha2.NewCOAError(nil, "unrecognized config type, accepted values are: path and bytes", v1alpha2.BadConfig)
+			sLog.Errorf("  P (K8s State): %+v", err)
+			return err
+		}
+	}
+	if err != nil {
+		sLog.Errorf("  P (K8s State): %+v", err)
+		return err
+	}
+	s.Clientset, err = kubernetes.NewForConfig(kConfig)
+	if err != nil {
+		sLog.Errorf("  P (K8s State): %+v", err)
+		return err
+	}
+
+	return nil
+}
+
+func toK8sStateProviderConfig(config providers.IProviderConfig) (K8sSecretProviderConfig, error) {
+	ret := K8sSecretProviderConfig{}
+	data, err := json.Marshal(config)
+	if err != nil {
+		return ret, err
+	}
+	err = json.Unmarshal(data, &ret)
+	if ret.ConfigType == "" {
+		ret.ConfigType = "path"
+	}
+	return ret, err
+}
+
+func (s *K8sSecretProvider) Read(name string, field string, localContext interface{}) (string, error) {
```

Comment:
Secret will be stored in based64 encoded form. Do we need to extend some configurations like decoding?

Reply:
the client would decode the secret